staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go

```diff
@@ -663,3 +663,346 @@ const (
+// +enum
+type PatchType string
+
+const (
+	// ApplyConfiguration indicates that the mutation is using apply configuration to mutate the object.
+	PatchTypeApplyConfiguration PatchType = "ApplyConfiguration"
+	// JSONPatch indicates that the object is mutated through JSON Patch.
+	PatchTypeJSONPatch PatchType = "JSONPatch"
+)
+
+// ApplyConfiguration defines the desired configuration values of an object.
+type ApplyConfiguration struct {
+	// expression will be evaluated by CEL to create an apply configuration.
+	// ref: https://github.com/google/cel-spec
+	//
+	// Apply configurations are declared in CEL using object initialization.  For example, this CEL expression
+	// returns an apply configuration to set a single field:
+	//
+	//	Object{
+	//	  spec: Object.spec{
+	//	    serviceAccountName: "example"
+	//	  }
+	//	}
+	//
+	// CEL expressions have access to the object types needed to create apply configurations:
+	// - 'Object' - CEL type of the object.
+	// - 'Object.<fieldName>' - CEL type of the object's spec (such as 'Object.spec')
+	// - 'Object.<fieldName1>.<fieldName2>...<fieldNameN>` - CEL type of a nested field (such as 'Object.spec.container')
+	//
+	// CEL expressions have access to the contents of the API request/response, organized into CEL variables as well as some other useful variables:
+	//
+	// - 'object' - The object from the incoming request. The value is null for DELETE requests.
+	// - 'oldObject' - The existing object. The value is null for CREATE requests.
+	// - 'request' - Attributes of the API request([ref](/pkg/apis/admission/types.go#AdmissionRequest)).
+	// - 'params' - Parameter resource referred to by the policy binding being evaluated. Only populated if the policy has a ParamKind.
+	// - 'namespaceObject' - The namespace object that the incoming object belongs to. The value is null for cluster-scoped resources.
+	// - 'variables' - Map of composited variables, from its name to its lazily evaluated value.
+	//   For example, a variable named 'foo' can be accessed as 'variables.foo'.
+	//
+	// The `apiVersion`, `kind`, `metadata.name` and `metadata.generateName` are always accessible from the root of the
+	// object. No other metadata properties are accessible.
+	//
+	// Only property names of the form `[a-zA-Z_.-/][a-zA-Z0-9_.-/]*` are accessible.
+	// Required.
+	Expression string `json:"expression,omitempty" protobuf:"bytes,1,opt,name=expression"`
+}
+
+// JSONPatch defines a JSON Patch.
+type JSONPatch []JSONPatchOperation
+
+// JSONPatchOperation defines a JSON Patch operation.
+type JSONPatchOperation struct {
+	// op specifies the operation this patch performs.
+	// Required.
+	Op JSONPatchOperationType `json:"op" protobuf:"bytes,1,opt,name=op"`
+
+	// path specifies the path to the target value.
+	// path must be a '/' delimited field path conforming to the [JSON Path spec](https://jsonpatch.com/).
+	// Required.
+	Path string `json:"path" protobuf:"bytes,2,opt,name=path"`
```

Comment:
I think this might be needed.

The digs into the bigger topic of json patch expressiveness, so  I should probably at least summarize two directional alternatives:

Option 1: Define the JSON Patch in the k8s API, with expression support for some fields (what I have in this PR now)

```
jsonPatch:
  - op: "add"
    path: "/metadat/label/x"
    valueExpression: "'y'"
```

Option 2: Have CEL return the entire JSON Patch:

```
jsonPatch:
  expression: `
     JSONPatch{
       [
         Op: "add",
         Path: "/metadata/labels/x",
         Value: "y"
       ]
     }
  `
````

If we continue with Option 1, we can add `patchExpression` and `fromExpression` to get some addtional expressiveness.

If we go with Option 2, it becomes possible to do things like use JSON Patch to modify all containers in a pod, which will never be possible with Option 1.

I'd admit I started building out Option 1 without seriously considering such use cases.  But they're on my mind.

ApplyConfigurations can update all container in a pod quite easily, but have limitations around removing values and updating atomic maps/lists/structs.

Reply:
I've switched `path` and `from` to CEL expressions (`pathExpression` and `fromExpression`).  Happy to discuss expressiveness topic further.